crypto/objects/obj_dat.c

```diff
@@ -856,10 +761,65 @@ const unsigned char *OBJ_get0_data(const ASN1_OBJECT *obj)
 
 int OBJ_add_object(const ASN1_OBJECT *obj)
 {
-    return ossl_obj_add_object(obj, 1);
+    ASN1_OBJECT *o = NULL;
+    ADDED_OBJ *ao[4] = { NULL, NULL, NULL, NULL }, *aop[4];
+    int i;
+
+    if ((o = OBJ_dup(obj)) == NULL)
+        return NID_undef;
+    if ((ao[ADDED_NID] = OPENSSL_malloc(sizeof(*ao[0]))) == NULL
+            || (o->length != 0
+                && obj->data != NULL
+                && (ao[ADDED_DATA] = OPENSSL_malloc(sizeof(*ao[0]))) == NULL)
+            || (o->sn != NULL
+                && (ao[ADDED_SNAME] = OPENSSL_malloc(sizeof(*ao[0]))) == NULL)
+            || (o->ln != NULL
+                && (ao[ADDED_LNAME] = OPENSSL_malloc(sizeof(*ao[0]))) == NULL))
+        goto err2;
+
+    if (!ossl_obj_write_lock(1)) {
+        ERR_raise(ERR_LIB_OBJ, ERR_R_UNABLE_TO_GET_WRITE_LOCK);
+        goto err2;
+    }
```

Comment:
I think you need to check again that it isn't added in the mean time, by calling the lh_ADDED_OBJ_retrieve() function.

Reply:
dang, I thought the hash table returned error on collision.  I really hate that our hash table requires two traversals  of the hash table to do a non-replacing insert.  I'm going to update it instead to allow for a single check.

Reply:
We do, its on line 791.

Though that would be a good improvement for the future.  It stinks that we need to do two lhash traversals for each addition where we want to avoid collisions with already added objects.

We should add an lh_TYPE_insert_noreplace method to the hashtable so as to reduce these to a single lookup.  I'll open an issue to track that